wrangler-service/src/main/java/co/cask/wrangler/service/common/AbstractWranglerService.java

```diff
@@ -21,32 +21,62 @@
 import co.cask.cdap.api.service.http.AbstractHttpServiceHandler;
 import co.cask.cdap.api.service.http.HttpServiceContext;
 import co.cask.wrangler.DataPrep;
+import co.cask.wrangler.PropertyIds;
 import co.cask.wrangler.dataset.connections.ConnectionStore;
 import co.cask.wrangler.dataset.workspace.DataType;
 import co.cask.wrangler.dataset.workspace.WorkspaceDataset;
 import co.cask.wrangler.dataset.workspace.WorkspaceException;
 import co.cask.wrangler.proto.Recipe;
 import co.cask.wrangler.proto.RequestV1;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
 import com.google.gson.Gson;
 
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
 import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
 import javax.annotation.Nullable;
 
+import org.apache.hadoop.security.UserGroupInformation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import static co.cask.wrangler.service.directive.DirectivesService.WORKSPACE_DATASET;
 
 public class AbstractWranglerService extends AbstractHttpServiceHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractHttpServiceHandler.class);
   private static final Gson GSON = new Gson();
   protected ConnectionStore store;
+  private static final long UGI_CACHE_EXPIRATION_MS = 3600000L;
+  private static LoadingCache<UGICacheKey, UserGroupInformation> ugiCache;
```

Comment:
I think key should be String (logged in user), you can make keytab and principal instance level variables. They are not gona change per call.

Reply:
Done. Thanks for feedback